ludwig/models/llm.py

```diff
@@ -418,23 +489,117 @@ def get_args(self):
+            for t in [GPT2Tokenizer, GPT2TokenizerFast, LlamaTokenizer, LlamaTokenizerFast]
+        ):
+            self.tokenizer.pad_token = self.tokenizer.eos_token
+            self.tokenizer.pad_token_id = self.tokenizer.eos_token_id
+
+    def _generate_merged_ids(self, input_ids, target_ids):
+        """This function merges the input_ids and target_ids together to create a unified tensor to pass into the
+        model.
+
+        This is required for PEFT based fine-tuning. It also returns attention masks for the merged tensors.
+        """
+
+        # target_ids is None during evaluation of the validation/test sets in the training loop.
+        if not torch.is_tensor(target_ids):
+            # Create attention masks for the input_ids.
+            attention_masks = []
+            for input_id_sample in input_ids:
+                attention_masks.append(self._create_attention_mask(input_id_sample))
+            return input_ids, torch.stack(attention_masks)
+
+        merged_input_and_targets = []
+        lengths = []
+
+        pad_tensor = torch.tensor([self.tokenizer.pad_token_id]).to(target_ids[0].device)
+
+        # Merge input_ids and target_ids by concatenating them together.
+        # We remove the left padding from both input_ids and target_ids before concatenating them.
+        for input_id_sample, target_id_sample in zip(input_ids, target_ids):
+            input_id_sample_no_padding = self._remove_left_padding(input_id_sample)[0]
+            target_id_sample_no_padding = torch.cat(
+                (self._remove_left_padding(target_id_sample)[0], pad_tensor), dim=-1
+            )
+
+            merged_sample_ids = torch.cat((input_id_sample_no_padding, target_id_sample_no_padding), dim=-1)
+
+            merged_input_and_targets.append(merged_sample_ids)
+            lengths.append(merged_sample_ids.shape[0])
+
+        # Since we remove the left padding from the target_ids, the merged input_ids and target_ids
+        # may not have the same lengths. We need to align them to the same length by adding left padding
+        # and generate an attention mask for just the part of the input that is not padding.
+        max_length = max(lengths)
+        attention_masks = []
+        for i, merged_sample_ids in enumerate(merged_input_and_targets):
+            merged_input_and_targets[i] = self._add_left_padding(merged_sample_ids, max_length)
+            attention_masks.append(self._create_attention_mask(merged_input_and_targets[i]))
+
+        return torch.stack(merged_input_and_targets), torch.stack(attention_masks)
+
     def _remove_left_padding(self, input_ids_sample: torch.Tensor):
-        bos_idxs = torch.where(input_ids_sample == self.tokenizer.bos_token_id)[0]  # all BOS token locations
+        """Removes left padding from the input_ids tensor."""
+        # Remove all PAD tokens
+        pad_idxs = torch.where(input_ids_sample == self.tokenizer.pad_token_id)[0]  # all PAD token locations
```

Comment:
What if there is no left padding because all of the sequences are the same length (or worse, they are right padded somehow)

Let's make sure we add some checks here (would also really like some unit tests to increase confidence in this since its quite complex)